File: opentsdb-proxy/src/main/java/net/opentsdb/proxy/netty/ConnectionManager.java
package net.opentsdb.proxy.netty;

import org.jboss.netty.channel.*;
import org.jboss.netty.channel.group.DefaultChannelGroup;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.channels.ClosedChannelException;

public class ConnectionManager extends SimpleChannelHandler {
  private static final Logger logger = LoggerFactory.getLogger(ConnectionManager.class);
  private static final DefaultChannelGroup channels = new DefaultChannelGroup("all-channels");

  static void closeAllConnections() {
    channels.close().awaitUninterruptibly();
  }

  /**
   * Constructor.
   */
  public ConnectionManager() {
  }

  @Override
  public void channelOpen(final ChannelHandlerContext ctx, final ChannelStateEvent e) {
    channels.add(e.getChannel());
  }

  @Override
  public void handleUpstream(final ChannelHandlerContext ctx, final ChannelEvent e) throws Exception {
    if (e instanceof ChannelStateEvent) {
      logger.info(e.toString());
    }
    super.handleUpstream(ctx, e);
  }

  @Override
  public void exceptionCaught(final ChannelHandlerContext ctx, final ExceptionEvent e) {
    final Throwable cause = e.getCause();
    final Channel chan = ctx.getChannel();
    if (cause instanceof ClosedChannelException) {
      logger.warn("Attempt to write to closed channel " + chan);
      return;
    }
    if (cause instanceof IOException) {
      final String message = cause.getMessage();
      if ("Connection reset by peer".equals(message)
          || "Connection timed out".equals(message)) {
        // Do nothing. A client disconnecting isn't really our problem. Oh,
        // and I'm not kidding you, there's no better way to detect ECONNRESET
        // in Java. Like, people have been bitching about errno for years,
        // and Java managed to do something *far* worse. That's quite a feat.
        return;
      }
    }
    logger.error("Unexpected exception from downstream for " + chan, cause);
    e.getChannel().close();
  }

}

File: opentsdb-client/src/main/scala/net/opentsdb/client/netty/ConnectionManager.scala
package net.opentsdb.client.netty

import org.jboss.netty.channel._
import org.slf4j.{LoggerFactory, Logger}
import org.jboss.netty.channel.group.DefaultChannelGroup
import java.nio.channels.ClosedChannelException
import java.io.IOException

object ConnectionManager {
  final val channels: DefaultChannelGroup = new DefaultChannelGroup("all-channels")
}

class ConnectionManager extends SimpleChannelHandler {
  private final val logger: Logger = LoggerFactory.getLogger(classOf[TsdbClient])

  override def channelOpen(ctx: ChannelHandlerContext, e: ChannelStateEvent) = ConnectionManager.channels.add(e.getChannel)

  override def handleUpstream(ctx: ChannelHandlerContext, e: ChannelEvent) = {
    e match {
      case e: ChannelStateEvent => logger.info(e.toString)
      case _ =>
    }

    super.handleUpstream(ctx, e)
  }

  override def exceptionCaught(ctx: ChannelHandlerContext, e: ExceptionEvent) = {
    val cause = e.getCause
    val chan = ctx.getChannel

    cause match {
      case causeMatch: ClosedChannelException => logger.warn("Attempting to write to a closed channel " + chan)
      case causeMatch: IOException => {
        val message = cause.getMessage
        if ("Connection reset by peer".equals(message) || "Connection timed out".equals(message)) {
          // Do nothing. A client disconnecting isn't really our problem. Oh,
          // and I'm not kidding you, there's no better way to detect ECONNRESET
          // in Java. Like, people have been bitching about errno for years,
          // and Java managed to do something *far* worse. That's quite a feat.
        }
      }
      case _ => {
        logger.error("Unexpected exception from downstream for " + chan, cause)
        e.getChannel.close()
      }
    }
  }
}
